### ISSUE TITLE
Clean test files

### ISSUE THREAD

**Issue by aicioara** [palantir/plottable#2504]
They are messy and have various degrees of nesting. We should reduce the nesting as much as possible first. 

Probably the messiest are `barPlotTests`

**Comment by aicioara** [comment]
Might be a duplicate as well.

**Comment by bluong** [comment]
I personally don't have too much against nesting since it has the capability of unit test grouping which is great from a readability standpoint.  However, I do agree that our usage is inconsistent

**Comment by aicioara** [comment]
> which is great from a readability standpoint

Disagree. Look at bar plot test, at the 3rd-4th nesting and see if you figure out where the variables come from and which `beforeEach()` initialized them. 

I agree though that the issue with that file is not the nesting but the number of LOC, but the nesting multiplies the messiness.

**Comment by aicioara** [comment]
Actually my concern is not nesting, but beforeEach() overuse. There should be at most 1 beforeEach() per it()

**Comment by aicioara** [comment]
Since I am fighting with testing right now, here is another argument.

If the beforeEach() does too much it is impossible to review it properly. In one of the tests I am looking at, the beforeEach() sets the accessor, then the data is set in the test itself and is improperly formatted:

```typescript
beforeEach(function() {
  // ...  
  plot.x(function(d) { return d.value }
  // ...
})

it("foo", function() {
  plot.addDataset(new Plottable.Dataset([{x : 2}]) // It should be value: 2
})
```

**Comment by bluong** [comment]
I near 100% agree that oftentimes our `beforeEach()` usage is mainly inconsistent, but this does not mean that `beforeEach()` is necessarily overused.

Here's the problem I found with `beforeEach()`:

The purpose of `beforeEach()` is to preconfigure an environment for some tests to run in.  This works great if the resultant environment is very well-defined, but the unfortunate situation in our unit tests is that what we want out of our resultant environment is not well-defined.

You see this all over the place when we have a `beforeEach()` for something like "basic bar plot" (a very not well-defined description) where we preconfigure the plot with a bunch of accessors, for example.

The next thing that happens is another developer will enter the scene with a unit test for a bar plot, in which he would look at this block and try to push his test there cause it's "a test for a basic bar plot".

The next thing that happens is, of course, he finds that the preconfigured accessors are exactly what he doesn't want.  Maybe also, the plot has an existing dataset that is not useful under his test.  All of a sudden you have a bunch of awkward code at the front of the unit test that is written specifically to get rid of the preconfiguration (the obvious follow up question is then "why did the preconfiguration happen in the firs place?").

The problem in the above is not necessarily the `beforeEach()` usage.  It is in the expression of the unit test suite.  Specifically, the name of the unit test suite did not holistically match the expression of the `beforeEach()` block, mainly in the fact that it was way too general / ambiguous / vague.

In which case, I can think of 2 solutions to the above scenario, depending on the intention of the author.  If the author wanted this test block to work with "a basic bar plot", the accessor configurations is definitely too much for the `beforeEach()` block and this should be done in the appropriate unit test.  If the author wanted this test block to actually be "a bar plot with the this type of accessor", then the accessor configurations I think are perfectly fine; he just needs to modify the name of the test block title to something more accurate.

**Comment by aicioara** [comment]
Agree, but I think you should summarize the above. I got lost in the words. If the guidelines are not straightforward and easy to follow, they will not be followed. 

What is the essence basically? That devs should pay more attention to what they write? That CRs should be more thorough? In my opinion, tests should be simpler, easier to read. That means less bugs. As a requirement, it needs less nesting, or shorter test files. 

Testing is our most precious resource. If we have bugs in our tests, we can't do anything about bugs in our code.

**Comment by bluong** [comment]
TLDR - The name of the unit test suite should be accurate with the representation of the environment after the `before`/`beforeEach` block.  If it is not accurate, developers will get confused and write tests in undesired places, and this can easily result in badly written tests.
